File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int count = 0;
        int currSum = 0;
        mp.put(0, 1);
        for (int num : nums) {
            currSum += num;
            if (mp.containsKey(currSum - goal)) {
                count += mp.get(currSum - goal);
            }
            mp.put(currSum, mp.getOrDefault(currSum, 0) + 1);
        }
        return count;
    }
}

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int n = nums.size();
        unordered_map<int, int> mp;
        int count   = 0;
        int currSum = 0;
        mp[0] = 1;
        for(int &num : nums) {
            currSum += num;
            if (mp.find(currSum - goal) != mp.end()) {
                count += mp[currSum - goal];
            }            
            mp[currSum]++;          
        }        
        return count;
    }
};